datafusion/core/src/physical_optimizer/repartition.rs

```diff
@@ -546,12 +538,12 @@ mod tests {
     }
 
     #[test]
-    fn repartition_ignores_sort_preserving_merge() -> Result<()> {
+    fn repartition_with_preserving_merge() -> Result<()> {
         let plan = sort_preserving_merge_exec(parquet_exec());
 
         let expected = &[
             "SortPreservingMergeExec: [c1@0 ASC]",
-            // Expect no repartition of SortPreservingMergeExec
+            "RepartitionExec: partitioning=RoundRobinBatch(10)",
```

Comment:
This seems worse than before? I think round robin should not be added before merge (at least currently, as we don't support parallel sort very well) as it will increase the work needed.

Reply:
Hi @Dandandan, actually, the UT here is not for the whole optimizing process, since the manually constructed plan is not a real case. The real case should insert a parallel local sort as the input of `SortPreservingMergeExec`, which is good for the round robin `RepartitionExec` between the `ParquetExec` and the `SortExec`.

What's more, current rule of `RepartitionExec` employs a bottom-up policy rather than the top-down policy, which makes it hard to consider whether to insert a `RepartitionExec` as an input of some operator.

Reply:
I am pretty sure these tests were added explicitly to avoid the case @Dandandan  is referring to.

Given that they tests are explicitly for "not putting the repartition exec" before a SortPreservingMerge I don't think we should simply change the tests.

If the issue is that the unit test doesn't accurately reflect the output of the whole optimizer process with all the rules, perhaps we can create a test that runs the whole physical optimizer  and move the "no repartition before the sort preserving merge test" ther?

Here is an example of such a test for the logical optimizer;  
https://github.com/apache/arrow-datafusion/blob/master/datafusion/optimizer/tests/integration-test.rs

I can try to help with this test framework, if it would help